### PR TITLE
W7 - TungNguyen - UnitTest

### PR DESCRIPTION
**Pull request by at-tungnguyen2**
#### Related issues (Optional)
Task #

#### Description (Required)
- [x] Create Username and Password Validation
- [x]  Create Unit test

### CODE REVIEW

File: app/src/testAt-tungnguyenDebug/java/vn/asiantech/internship/tungnguyen/PasswordUnitTest.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ * Created by tungnguyen on 03/01/2018.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class PasswordUnitTest {
+    @Mock
+    private User mUser;
+
+    @Test
+    public void isCheckPasswordDifferentUser() {
+        Mockito.when(mUser.getUserName()).thenReturn("Tungkute");
```
Comment (at-dev-support-01) [inline app/src/testAt-tungnguyenDebug/java/vn/asiantech/internship/tungnguyen/PasswordUnitTest.java]: WARNING: [AvoidDuplicateLiterals] The String literal "Tungkute" appears 4 times in this file; the first occurrence is on line 25
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/strings.html#AvoidDuplicateLiterals

**Comment by at-tungnguyen2** [comment]
@at-trungnguyen  Please check for me, thank you very much.

File: app/src/at-tungnguyen/java/vn/asiantech/internship/ui/validation/UserValidation.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+import java.util.Locale;
+import java.util.regex.Pattern;
+
+/**
+ * Author Asian Tech Inc.
+ * Created by tungnguyen on 03/01/2018.
+ */
+
+public class UserValidation {
```
Comment (at-trungnguyen) [inline app/src/at-tungnguyen/java/vn/asiantech/internship/ui/validation/UserValidation.java]: fix same class above, use `final`

File: app/src/at-tungnguyen/java/vn/asiantech/internship/ui/validation/UserValidation.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    public UserValidation() {
```
Comment (at-trungnguyen) [inline app/src/at-tungnguyen/java/vn/asiantech/internship/ui/validation/UserValidation.java]: `private`

File: app/src/at-tungnguyen/java/vn/asiantech/internship/ui/validation/UserValidation.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+     */
+    public static boolean isCheckUserCapitalAndNumber(String userName) {
+        int numCount;
+        boolean capitalCharacter = false;
+        boolean numberCharacter = false;
+        for (int i = 0; i < userName.length(); i++) {
+            if (Character.isUpperCase(userName.charAt(i))) {
+                capitalCharacter = true;
+            } else if (Character.isDigit(userName.charAt(i))) {
+                numCount = i;
```
Comment (at-trungnguyen) [inline app/src/at-tungnguyen/java/vn/asiantech/internship/ui/validation/UserValidation.java]: I think it's wrong 💯

File: app/src/at-tungnguyen/res/layout/activity_login3.xml [hunk 5]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (at-trungnguyen) [inline app/src/at-tungnguyen/res/layout/activity_login3.xml]: format this layout

File: app/src/at-tungnguyen/res/values/strings.xml [hunk 6]
```diff
@@ -81,5 +81,35 @@
...
+    <string name="btn_thread_ex2">Exercise2</string>
+    <string name="tv_count">0</string>
+    <string name="tv_percent">0%</string>
+    <string name="have_not_picked_img">Not picked</string>
+    <string name="tv_countdown_timer">Tungkute</string>
+    <string name="tv_username">UserName</string>
+    <string name="tv_password">Password</string>
+    <string name="tv_test_login"> Login </string>
+    <string name="btn_login_validation">Validation</string>
+    <string name="user_name_l">Độ dài username lớn hơn 5 và nhỏ hơn 24 kí tự</string>
```
Comment (at-trungnguyen) [inline app/src/at-tungnguyen/res/values/strings.xml]: use EL.
fix all

File: app/src/testAt-tungnguyenDebug/java/vn/asiantech/internship/tungnguyen/PasswordUnitTest.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        Mockito.when(mUser.getUserName()).thenReturn("Tunghandsome");
+        Mockito.when(mUser.getPassword()).thenReturn("TunghandSome");
+        Assert.assertTrue(PasswordValidation.isCheckPasswordDifferentUser(mUser.getUserName(), mUser.getPassword()));
+    }
+
+    @Test
+    public void isCheckPassSpecial() {
+        Mockito.when(mUser.getPassword()).thenReturn("tung");
+        Assert.assertTrue(PasswordValidation.isCheckPassSpecial(mUser.getPassword()));
+//        Mockito.when(mUser.getPassword()).thenReturn("tungmoutian");
```
Comment (at-trungnguyen) [inline app/src/testAt-tungnguyenDebug/java/vn/asiantech/internship/tungnguyen/PasswordUnitTest.java]: delete it

File: app/src/testAt-tungnguyenDebug/java/vn/asiantech/internship/tungnguyen/PasswordUnitTest.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+//        Assert.assertFalse(PasswordValidation.isCheckPassSpecial(mUser.getPassword()));
```
Comment (at-trungnguyen) [inline app/src/testAt-tungnguyenDebug/java/vn/asiantech/internship/tungnguyen/PasswordUnitTest.java]: delete it

File: app/src/testAt-tungnguyenDebug/java/vn/asiantech/internship/tungnguyen/PasswordUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+ * Author Asian Tech Inc.
+ * Created by tungnguyen on 03/01/2018.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class PasswordUnitTest {
+    @Mock
+    private User mUser;
+
+    @Test
+    public void isCheckPasswordDifferentUser() {
```
Comment (at-trungnguyen) [inline app/src/testAt-tungnguyenDebug/java/vn/asiantech/internship/tungnguyen/PasswordUnitTest.java]: `checkPasswordDifferentUser()` I think 👎

File: app/src/testAt-tungnguyenDebug/java/vn/asiantech/internship/tungnguyen/PasswordUnitTest.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        Mockito.when(mUser.getUserName()).thenReturn("Tungkute");
+        Mockito.when(mUser.getPassword()).thenReturn("Tungkute");
+        Assert.assertFalse(PasswordValidation.isCheckPasswordDifferentUser(mUser.getUserName(), mUser.getPassword()));
+        Mockito.when(mUser.getUserName()).thenReturn("Tunghandsome");
+        Mockito.when(mUser.getPassword()).thenReturn("TunghandSome");
+        Assert.assertTrue(PasswordValidation.isCheckPasswordDifferentUser(mUser.getUserName(), mUser.getPassword()));
+    }
+
+    @Test
+    public void isCheckPassSpecial() {
```
Comment (at-trungnguyen) [inline app/src/testAt-tungnguyenDebug/java/vn/asiantech/internship/tungnguyen/PasswordUnitTest.java]: fix it same above

File: app/src/testAt-tungnguyenDebug/java/vn/asiantech/internship/tungnguyen/PasswordUnitTest.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+import vn.asiantech.internship.model.User;
+import vn.asiantech.internship.ui.validation.PasswordValidation;
+
+
+/**
+ * Author Asian Tech Inc.
+ * Created by tungnguyen on 03/01/2018.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class PasswordUnitTest {
```
Comment (at-trungnguyen) [inline app/src/testAt-tungnguyenDebug/java/vn/asiantech/internship/tungnguyen/PasswordUnitTest.java]: enter 1 line

File: app/src/testAt-tungnguyenDebug/java/vn/asiantech/internship/tungnguyen/UserUnitTest.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+ * Created by tungnguyen on 04/01/2018.
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class UserUnitTest {
+
+    @Mock
+    private User mUser;
+
+    @Test
+    public void ischeckUserLenght() {
```
Comment (at-trungnguyen) [inline app/src/testAt-tungnguyenDebug/java/vn/asiantech/internship/tungnguyen/UserUnitTest.java]: fix same above, `checkUserLenght()`

File: app/src/testAt-tungnguyenDebug/java/vn/asiantech/internship/tungnguyen/UserUnitTest.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        Mockito.when(mUser.getUserName()).thenReturn("abcabc");
+        Assert.assertTrue(UserValidation.ischeckUserLenght(mUser.getUserName()));
+        Mockito.when(mUser.getUserName()).thenReturn("abcab");
+        Assert.assertFalse(UserValidation.ischeckUserLenght(mUser.getUserName()));
+    }
+
+    @Test
+    public void isCheckUserSpecial() {
```
Comment (at-trungnguyen) [inline app/src/testAt-tungnguyenDebug/java/vn/asiantech/internship/tungnguyen/UserUnitTest.java]: fix all below

**Comment by at-tungnguyen2** [comment]
@at-trungnguyen  Please review for me, tks you

**Pull request merged**